File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

    
def solution(begin, target, words):
    answer = 0
    q = deque()
    q.append((begin,0))
    # 단어 방문 
    visited = set()
    while q :
        w,idx = q.popleft()
        # 현재 단어랑 target 이랑 같으면 return idx 
        if w == target :
            return idx
        for word in words :
            # 방문하지 않고 한글자만 다르면 q 에 삽입
            if word not in visited and sum([w[i] != word[i] for i in range(len(word))]) == 1 :
                q.append((word,idx+1))
                visited.add(word)
        
    return answer

File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
    var answer = 0;
    
    if(!words.includes(target)) {
        return 0;
    }
    
    let visited = new Array(words.length).fill(0);   
    
    let queue = [[begin,0]];
    
    while(queue.length > 0) {
        let [currentWord,changeCount] = queue.shift();
        if(currentWord === target) {
            return changeCount;
        }
        
        words.forEach((word,index) => {
            let diff = 0;
            
            if(!visited[index]) {
                for(let i=0;i<currentWord.length;i++){
                    if(currentWord[i] !== word[i]) {
                        diff++;
                    }
                    
                }
            }
            if(diff === 1){
                queue.push([word,changeCount+1]);
                visited[index]= 1;
            }
        })
    }
    return 0;
}